### ISSUE TITLE
Support importing animated GIFs

### ISSUE THREAD

**Issue by XDelta** [Yellow-Dog-Man/Resonite-Issues#1266]
### Is your feature request related to a problem? Please describe.

Currently importing a gif will import a single frame as a StaticTexture2D.

### Describe the solution you'd like

I'd like to see the ability for gifs to be imported in a form that is animated.

### Describe alternatives you've considered

Converting gif images into a sprite atlas and animating UV's however this only really works for lower framerate, short duration, and/or low resolution gifs. Otherwise the texture becomes very large.

Online tools exist to make sprite sheets from gif's https://ezgif.com/gif-to-sprite or converting them to another format such as mp4.
A mod also exists for converting them and setting up the components for a basic animation https://github.com/astralchan/GifImporter (although it recently broke).

**Comment by Frooxius** [comment]
I've looked into this in the past, but this is a bit of a tricky issue. One approach would be to treat GIF's as video assets, with a timeline, that you can pause, resume, seek around like any video. 

However GIF's don't really fit into that format super neatly. They support complex looping - like parts of the GIF might loop around. But also they support various ways of replacing previous frames, including replacing only parts of it. There are some GIFs that use that to create various effects by stacking frames on top of each other.

This makes stuff like seeking more difficult, because to get the correct visual at the time, you'd need to process the whole playback up to that point - which can be computationally pretty expensive.

Even with simple ones that do replace frames, each frame can take a different amount of time to be displayed - so this needs some handling to figure out which frame to show more efficiently.

So we have a few choices here:
- Treat GIF's as a "stream" - they either play or don't, but we don't guarantee they stay synchronized between players - each one might see them at different point in time
- Treat them as video with timeline - this would probably work for majority of GIFs that people want to bring in and would ensure synced playback, but would probably do weird things with the more "special" GIFs

Implementing either will take a chunk of time though. There's a few approaches here too:
- Bundle a 3rd party tool that converts GIFs into videos on import and that's then imported as normal video - this would be a lot more easier to implement - it's essentially a fancier version of that conversion to atlas
- Support decoding and playing them back natively, which would be a bit more seamless/performant and would support the "weirder" GIFs, but also a lot more work

So that brings last question - is either of those worth the engineering effort?

Nowadays a lot of things are handled as videos and I don't know how often you see animated GIFs anymore and what would be the usecase for these that you can't achieve through other means or just converting them to video yourself.

How often do people need to import GIFs? How much of an obstacle is this? More info on this would help us gauge if it's worth prioritizing.

**Comment by XDelta** [comment]
At least in my opinion, the 'weird' variants of gif's come up so infrequently, if at all, without trying to find them specifically it would probably be more reasonable to go with a simpler implementation that is similar to videos.

**Comment by Nutcake** [comment]
Would it make sense to add support for animated WebPs at the same time? (Assuming they're not already supported, I cannot check right now) 

As for if it is worth it, I'd say it is, considering the currently largest chat platform (discord) as well as one of the largest social media sites (reddit) both have features that allow direct searching for and embedding of gifs in their respective apps which users may want to share in Resonite. GIFs are unfortunately still a very prevalent format and I'd say supporting them is pretty essential.

**Comment by jae1911** [comment]
Also roughly related to #880 since AVIF supports sequences as well.

**Comment by XDelta** [comment]
If it had a mechanism that could convert them on import to an already supported video format, I think that would cover most 'wants/needs' of supporting gif as a format. At least in my experience, it pretty much always is browsing some site and copying a gif in to share to find that it doesn't work or rather is a single frame.

**Comment by JackTheFoxOtter** [comment]
I agree. I think what people want with gif support is the ability to drop in a gif file and display it in the world. Converting them to videos would fulfill that request. Native GIF format support would be a fun novelty feature, but I don't think it would actually bring many real-world benefits.

**Comment by Frooxius** [comment]
Yeah, we can go that route if people are interested. Essentially a convenience option for converting those assets. 

One more caveat with that though is... do people care about transparent GIFs with this? Is alpha important? That might add some more constraints.

**Comment by JackTheFoxOtter** [comment]
I mean, I don't think the default video player would be see through if a video is transparent. regardless of if it was a gif or not. So I don't think transparency support would be super important.

We'd also have the playback controls of the player, so I imagine it would look quite weird being transparent.

**Comment by VirgilCore** [comment]
Tranparency is important when using the gif as an asset for a material or such
Videos support transparency so why shouldn't gifs?

**Comment by TisFoolish** [comment]
A LOT of pixel art is animated and in the gif format. Whatever method is chosen, I hope it allows for point texture filtering. Anything else would ruin the art

**Comment by Frooxius** [comment]
That might actually be a good argument for more native support. You could do point filtered video, but I think the video compression would hurt it too.

**Comment by TisFoolish** [comment]
Slightly offtopic, but wasn't Geenz working on that and the audio desync issue the video player has?

**Comment by Frooxius** [comment]
Please keep this on topic. This is for adding support for GIF, not really other issues with existing video playback.

**Comment by sveken** [comment]
It comes up enough that there was a mod created for this. 
The way the mod gets around it to my understanding is it converts the GIF into a sprite sheet then drives the frame with a timer. Thus turning the sprite sheet back into a scrolling gif. 

Instead of making it a video, would it be easier to just turn each gif frame into an image (or even a sprite sheet) instead and just drive the frame playback?

**Comment by Frooxius** [comment]
That only works for short GIFs. If you bring big ones, it would end up using a lot of VRAM.

**Comment by iamgreaser** [comment]
> They support complex looping - like parts of the GIF might loop around.

As far as I can tell, they don't. However, due to how disposal methods could be misused, it may be necessary to run through the animation twice... assuming that such an approach is actually supported by anything, if not then it should be adequate to only run through the animation once.

----

Anyway, let's cover all the forms of weirdness one can encounter with an animated GIF. I am basing this on the GIF89a specification, plus a little bit I gleaned from Wikipedia on the Netscape extension block. To clarify: All animated GIFs MUST be GIF89a - unanimated GIFs can be GIF87a which is fully backwards-compatible with GIF89a.

Even without the animated part of animated GIF, each image part can have any of the following:

* Its own position and bounds. This is par for the course, but does bring its own funkiness.
* Its own local palette. This is probably the nastiest case, as it prevents operating entirely in the 8 bits per pixel realm.
* A flag to indicate that said image part is to be stored interlaced. Not nasty, just a bit annoying to need to deal with.

The "NETSCAPE2.0" extension block is used to indicate that a GIF is animated. The only field in it allows one to indicate the number of times the GIF should be animated. Personally I would be tempted to ignore the value of this field and have it keep looping, as attempting to limit the number of times it can play will add complexity.

Most of the animation magic takes place in the Graphic Control Extension (GCE) block.

Each GCE block has an optional transparent colour index to indicate which pixel is to be treated as transparent - another reason why keeping it 8bpp is not always an option!

There are 3 actual disposal modes, as well as 1 "don't care" mode - it is up to the decoder as to which disposal method would work better, and if it actually matters which disposal method is used in this case then the image file is fundamentally wrong. The disposal modes are:

* Do not dispose - use the data from the previous frame.
* Restore to background colour - fill the given rectangle with the background colour.
* Restore to previous - replace the given rectangle with what was there one frame prior to what's currently there. This one's probably quite rare to encounter these days and the spec suggests that one restore it to the background colour if this is infeasible.

As for actual animation, there is a delay time to indicate how long to display a frame for. If this is 0, then the behaviour is somewhat undefined, but should probably be treated as 100 ms to make it somewhat consistent with browsers.

And lastly, the weirdest piece of weirdness, which you're probably going to ignore: **There is a flag to indicate that user input is expected before advancing to the next frame.**

----

I was about to make a proposal for how one could make this work for video but struggled to find a suitable codec - the DOSBox Zip Motion Blocks Video (ZMBV) codec came to mind but seems to be too dependent on delta frames and probably compresses worse than keyframe-only APNG... y'know what, I'm kinda tempted to make a codec for this case.

> That only works for short GIFs. If you bring big ones, it would end up using a lot of VRAM.

But that's the thing: It *works* for short GIFs. Covering this case in an import dialogue and allowing a user to choose would work quite nicely. The only place where this theoretically fails is in terms of delay times... can that be managed via one of the SomethingGradientDriver components?

Although admittedly in a lot of those cases, GIF is not the ideal format, and it would be better to directly import video as a sprite sheet...

**Comment by Frooxius** [comment]
We probably won't deal with most of these details themselves on the actual decoding - that'd be handled by library and we will just end up decoding a bitmap for display at each stage.

However what I was trying to get at, we can't just treat it as a simple video either - that's the details we do care about here.

Particularly the stuff like "do not dispose" makes it harder to do stuff like seeking. And similarly frames with variable length, that requires some handling too potentially, depending on how it's handled.

> But that's the thing: It works for short GIFs. Covering this case in an import dialogue and allowing a user to choose would work quite nicely. The only place where this theoretically fails is in terms of delay times... can that be managed via one of the SomethingGradientDriver components?

Yes, but that's the whole problem - it works _only_ for short GIFs. This makes this solution unacceptable as general GIF solution and not something we'd invest time into.

We could set arbitrary limit on size + length. People try to bring in bigger ones and fail - make an issue and we're back to square one. Because of that that's not really worth pursuing as an approach.

**Comment by jonnyawsom3** [comment]
With the mention of WebP and AVIF, APNG seems like the ideal replacement. Still lossless, still transparent, better compression and falls back to a normal image if unsupported.

**Comment by Frooxius** [comment]
How would APNG be a good replacement for this?

In my experience, it's a lot less commonly used than GIF online for animated images, so adding support for it wouldn't help people bring animated GIFs.

**Comment by iamgreaser** [comment]
APNG is a good replacement for GIF, and a major reason for its slow adoption was G :camera: :camera: gle was insisting on not supporting anything that stock libpng wouldn't support, and the people behind libpng refused to support it, until eventually someone found a way to massage libpng into supporting it or something, not sure what happened. That, and MNG (the format the PNG group have always supported) really is a complicated mess in comparison with PNG, to the point where you have to get to the level of WebP in order to make MNG look sane.

But if GIF or at least a sane subset of GIF cannot be supported for this purpose, then APNG won't magically be adequate either.

**Comment by Frooxius** [comment]
I still don't see how APNG would be good replacement in this particular case though?

The history of it is largely irrelevant, I am only looking through the lens of what benefits would it provide for Resonite and the use-cases outlined in this issue.

**Comment by Apollo1221alt** [comment]
> That only works for short GIFs. If you bring big ones, it would end up using a lot of VRAM.

In terms of VRAM issues and usage, that's why we could have a setting in the settings to turn on and off GIFs over a certain size, or when your not actively looking at it. On Discord for example there is a setting for GIFs, Strickers, and Emotes that requires clicking on them to playback the animations, just like a video player. For those who do not want to see the GIF we could impliment that, too? That would help performance anyway.

**Comment by Lunaero1** [comment]
> > That only works for short GIFs. If you bring big ones, it would end up using a lot of VRAM.
> 
> In terms of VRAM issues and usage, that's why we could have a setting in the settings to turn on and off GIFs over a certain size, or when your not actively looking at it. On Discord for example there is a setting for GIFs, Strickers, and Emotes that requires clicking on them to playback the animations, just like a video player. For those who do not want to see the GIF we could impliment that, too? That would help performance anyway.

commenting from my main cos i didnt realize i was on my other account when i posted this response, but are GIFs even vram hungry on browsers??

**Comment by Frooxius** [comment]
> > That only works for short GIFs. If you bring big ones, it would end up using a lot of VRAM.
> 
> In terms of VRAM issues and usage, that's why we could have a setting in the settings to turn on and off GIFs over a certain size, or when your not actively looking at it. On Discord for example there is a setting for GIFs, Strickers, and Emotes that requires clicking on them to playback the animations, just like a video player. For those who do not want to see the GIF we could impliment that, too? That would help performance anyway.

That only makes sense if you're actually decoding them on the fly (which is what Discord & web browsers do) - in which case you don't have the VRAM issue, because you're only keeping one frame in memory at the time (maybe two for double buffering). The reason Discord and similar do this isn't to conserve VRAM, it's to conserve CPU with playback.

It also doesn't fit with Resonite's system where things like this stay synced (e.g. consider video playback being synchronized).

The problem with a lot of VRAM is for approach that completely decodes GIFs into an atlas - in which case it doesn't matter for VRAM usage if they're playing or not - it was all decoded in advanced.
